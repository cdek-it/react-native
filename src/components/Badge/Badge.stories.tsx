import type { Meta, StoryObj } from '@storybook/react'
import React from 'react'
import { View } from 'react-native'

import { Badge } from './Badge'

const meta: Meta<typeof Badge> = {
  title: 'Badge',
  component: Badge,
  args: {
    severity: 'basic',
  },
  argTypes: {
    severity: { control: 'radio', options: ['basic', 'info', 'success', 'warning', 'danger'] },
  },
  decorators: [
    (Story) => (
      <View style={{ padding: 16 }}>
        <Story />
      </View>
    ),
  ],
}

export default meta

type Story = StoryObj<typeof Badge>

export const BadgeDot: Story = {
  args: {
    dot: true,
  },
  argTypes: {},
}

export const BadgeNormal: Story = {
  args: {
    children: 'Test',
  },
  argTypes: {},
}
