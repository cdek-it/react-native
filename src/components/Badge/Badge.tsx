import React from 'react'
import { Text, View, type ViewStyle } from 'react-native'

import { makeStyles } from '../../utils/makeStyles'

interface BadgeBase {
  severity?: 'basic' | 'info' | 'success' | 'warning' | 'danger'
  style?: ViewStyle
}

interface BadgeText extends BadgeBase {
  children: string
  dot?: never
}

interface BadgeDot extends BadgeBase {
  dot: true
  children?: never
}

export type BadgeProps = BadgeText | BadgeDot

export const Badge = ({ children, dot, severity = 'basic', style }: BadgeProps) => {
  const styles = useStyles()

  return (
    <View style={[styles.container, style]}>
      {dot ? (
        <View style={[styles.dot, styles[severity]]} />
      ) : (
        <Text style={[styles.textBadge, styles[severity]]}>{children}</Text>
      )}
    </View>
  )
}

const useStyles = makeStyles(({ theme }) => ({
  container: {
    alignItems: 'flex-start',
  },
  dot: {
    width: theme.Misc.Badge.badgeDotSize,
    height: theme.Misc.Badge.badgeDotSize,
    borderRadius: 100,
  },
  textBadge: {
    height: theme.Misc.Badge.badgeHeight,
    paddingHorizontal: theme.Misc.Tag.tagPadding,
    paddingVertical: theme.Misc.Chip.chipPaddingTopBottom,
    borderRadius: 100,
    verticalAlign: 'middle',
    fontFamily: 'Roboto',
    color: theme.Misc.Badge.badgeTextColor,
    fontSize: 10.5,
    fontWeight: 700,
  },
  basic: {
    backgroundColor: theme.Misc.Badge.badgeBg,
  },
  info: {
    backgroundColor: theme.Button.Severity.Info.Basic.infoButtonBg,
  },
  success: {
    backgroundColor: theme.Button.Severity.Success.Basic.successButtonBg,
  },
  warning: {
    backgroundColor: theme.Button.Severity.Warning.Basic.warningButtonBg,
  },
  danger: {
    backgroundColor: theme.Button.Severity.Danger.Basic.dangerButtonBg,
  },
}))
