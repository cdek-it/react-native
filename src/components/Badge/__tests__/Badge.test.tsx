import { render } from '@testing-library/react-native'
import React from 'react'

import { Badge, type BadgeProps } from '../Badge'

describe('Badge component tests', () => {
  const snapshotTestsConfig: BadgeProps[] = [
    { dot: true, severity: 'basic' },
    { dot: true, severity: 'info' },
    { dot: true, severity: 'success' },
    { dot: true, severity: 'warning' },
    { dot: true, severity: 'danger' },
    { children: '12', severity: 'basic' },
    { children: '12', severity: 'info' },
    { children: '12', severity: 'success' },
    { children: '12', severity: 'warning' },
    { children: '12', severity: 'danger' },
  ]

  snapshotTestsConfig.map((props) =>
    it(`Badge snapshot test: ${props.dot ? 'dot ' : ''} - ${props.severity}`, () => {
      const renderedBadge = render(<Badge {...props} />)
      expect(renderedBadge.toJSON()).toMatchSnapshot()
    })
  )

  it('Badge snapshot test: Badge with custom style', () => {
    const renderedBadge = render(<Badge dot severity='basic' style={{ margin: 10 }} />)
    expect(renderedBadge.toJSON()).toMatchSnapshot()
  })
})
