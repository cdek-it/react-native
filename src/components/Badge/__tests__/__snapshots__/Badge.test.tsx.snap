// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge component tests dot, severity: basic 1`] = `
<View
  style={
    [
      {
        "alignItems": "flex-start",
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "borderRadius": 100,
          "height": 7,
          "width": 7,
        },
        {
          "backgroundColor": "#188700",
        },
      ]
    }
  />
</View>
`;

exports[`Badge component tests dot, severity: danger 1`] = `
<View
  style={
    [
      {
        "alignItems": "flex-start",
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "borderRadius": 100,
          "height": 7,
          "width": 7,
        },
        {
          "backgroundColor": "#e26e63",
        },
      ]
    }
  />
</View>
`;

exports[`Badge component tests dot, severity: info 1`] = `
<View
  style={
    [
      {
        "alignItems": "flex-start",
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "borderRadius": 100,
          "height": 7,
          "width": 7,
        },
        {
          "backgroundColor": "#49a2ef",
        },
      ]
    }
  />
</View>
`;

exports[`Badge component tests dot, severity: success 1`] = `
<View
  style={
    [
      {
        "alignItems": "flex-start",
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "borderRadius": 100,
          "height": 7,
          "width": 7,
        },
        {
          "backgroundColor": "#68c170",
        },
      ]
    }
  />
</View>
`;

exports[`Badge component tests dot, severity: warning 1`] = `
<View
  style={
    [
      {
        "alignItems": "flex-start",
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "borderRadius": 100,
          "height": 7,
          "width": 7,
        },
        {
          "backgroundColor": "#ffb330",
        },
      ]
    }
  />
</View>
`;

exports[`Badge component tests severity: basic 1`] = `
<View
  style={
    [
      {
        "alignItems": "flex-start",
      },
      undefined,
    ]
  }
>
  <Text
    style={
      [
        {
          "borderRadius": 100,
          "color": "#ffffff",
          "fontFamily": "Roboto-Bold",
          "fontSize": 10.5,
          "height": 21,
          "paddingHorizontal": 7,
          "paddingVertical": 0,
          "verticalAlign": "middle",
        },
        {
          "backgroundColor": "#188700",
        },
      ]
    }
  >
    12
  </Text>
</View>
`;

exports[`Badge component tests severity: basic 2`] = `
<View
  style={
    [
      {
        "alignItems": "flex-start",
      },
      undefined,
    ]
  }
>
  <Text
    style={
      [
        {
          "borderRadius": 100,
          "color": "#ffffff",
          "fontFamily": "Roboto-Bold",
          "fontSize": 10.5,
          "height": 21,
          "paddingHorizontal": 7,
          "paddingVertical": 0,
          "verticalAlign": "middle",
        },
        {
          "backgroundColor": "#49a2ef",
        },
      ]
    }
  >
    12
  </Text>
</View>
`;

exports[`Badge component tests severity: basic 3`] = `
<View
  style={
    [
      {
        "alignItems": "flex-start",
      },
      undefined,
    ]
  }
>
  <Text
    style={
      [
        {
          "borderRadius": 100,
          "color": "#ffffff",
          "fontFamily": "Roboto-Bold",
          "fontSize": 10.5,
          "height": 21,
          "paddingHorizontal": 7,
          "paddingVertical": 0,
          "verticalAlign": "middle",
        },
        {
          "backgroundColor": "#68c170",
        },
      ]
    }
  >
    12
  </Text>
</View>
`;

exports[`Badge component tests severity: basic 4`] = `
<View
  style={
    [
      {
        "alignItems": "flex-start",
      },
      undefined,
    ]
  }
>
  <Text
    style={
      [
        {
          "borderRadius": 100,
          "color": "#ffffff",
          "fontFamily": "Roboto-Bold",
          "fontSize": 10.5,
          "height": 21,
          "paddingHorizontal": 7,
          "paddingVertical": 0,
          "verticalAlign": "middle",
        },
        {
          "backgroundColor": "#ffb330",
        },
      ]
    }
  >
    12
  </Text>
</View>
`;

exports[`Badge component tests severity: basic 5`] = `
<View
  style={
    [
      {
        "alignItems": "flex-start",
      },
      undefined,
    ]
  }
>
  <Text
    style={
      [
        {
          "borderRadius": 100,
          "color": "#ffffff",
          "fontFamily": "Roboto-Bold",
          "fontSize": 10.5,
          "height": 21,
          "paddingHorizontal": 7,
          "paddingVertical": 0,
          "verticalAlign": "middle",
        },
        {
          "backgroundColor": "#e26e63",
        },
      ]
    }
  >
    12
  </Text>
</View>
`;

exports[`Badge component tests with custom style 1`] = `
<View
  style={
    [
      {
        "alignItems": "flex-start",
      },
      {
        "margin": 10,
      },
    ]
  }
>
  <View
    style={
      [
        {
          "borderRadius": 100,
          "height": 7,
          "width": 7,
        },
        {
          "backgroundColor": "#188700",
        },
      ]
    }
  />
</View>
`;
